src/Server/TCPHandler.cpp

```diff
@@ -779,6 +782,8 @@ void TCPHandler::extractConnectionSettingsFromContext(const ContextPtr & context
 
 bool TCPHandler::readDataNext()
 {
+    std::scoped_lock lock(in_mutex);
```

Comment:
The linked data race report happens with `DB::TCPHandler::receiveReadTaskResponseAssumeLocked`, it does not go through `readDataNext`, but uses `in` directly, so this will not fix the issue.

Reply:
yes, missed this nuance